compiler/rustc_parse/src/parser/item.rs

```diff
@@ -71,20 +73,50 @@ impl<'a> Parser<'a> {
             if !self.maybe_consume_incorrect_semicolon(items.last().map(|x| &**x)) {
-                let msg = format!("expected item, found {token_str}");
-                let mut err = self.dcx().struct_span_err(self.token.span, msg);
-                let span = self.token.span;
-                if self.is_kw_followed_by_ident(kw::Let) {
-                    err.span_label(
-                        span,
-                        "consider using `const` or `static` instead of `let` for global variables",
-                    );
-                } else {
-                    err.span_label(span, "expected item")
-                        .note("for a full list of items that can appear in modules, see <https://doc.rust-lang.org/reference/items.html>");
+                let token_str = super::token_descr(&self.token);
+                let mut err = self
+                    .dcx()
+                    .struct_span_err(self.token.span, format!("expected item, found {token_str}"));
+
+                match self.parse_full_stmt(AttemptLocalParseRecovery::No) {
+                    Ok(Some(stmt)) => {
+                        let span = stmt.span;
+                        match stmt.kind {
+                            StmtKind::Let(_) => {
+                                err.span_label(span, "unexpected `let` binding outside of a function")
+                                    .help("consider using `const` or `static` instead of `let` for global variables");
+                            }
+
+                            StmtKind::Semi(e) | StmtKind::Expr(e) => {
+                                err.span_label(span, "unexpected expression").help(format!(
+                                    "consider putting it inside a function: fn main() {{ {} }}",
+                                    pprust::expr_to_string(&*e)
+                                ));
+                            }
+                            StmtKind::MacCall(_) => {
+                                err.span_label(span, "unexpected macro call at this position");
+                            }
+                            StmtKind::Item(_) => {
+                                err.span_label(span, "unexpected item at this position");
+                            }
+                            StmtKind::Empty => {
+                                unreachable!(
+                                    "should have been handled by maybe_consume_incorrect_semicolon"
+                                );
+                            }
+                        };
+                    }
+                    Err(err) => {
+                        // We don't really care about an error parsing this expression.
+                        err.cancel();
+                    }
+                    _ => {}
                 };
+                err.note("for a full list of items that can appear in modules (outside of a function), see <https://doc.rust-lang.org/reference/items.html>");
```

Comment:
The `function` is inaccurate, there are also scenarios such as `method`, `closure`, `macro` and so on. I recommend removing the related tip.

Reply:
I don't think many people actually know what an item is, and the people who do don't need the note. I think it can be worth clarifying it.
By the same logic I think it's worth avoiding words like "token" in error messages. People that have never done lexing probably don't know what a token is.

I've opened up [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/We.20should.20avoid.20words.20like.20token.20and.20item.20in.20the.20diagnostics).